test across multiple branches

smalls opened the pull request.

smalls committed:
package index.test.html with the appropriate branch
Currently there are only tests for v0.2.1, so that's the only version
with a test.

smalls committed:
'npm test' should run tests in multiple branches
Specifically, all mocha-chrome tests matching */test/index.test.html.
This is targeted at Travis/Linux, it may not work on Windows since I'm
depending on a filename glob in package.json. If we need Windows I'd
probably recommend pushing that logic from a glob into
mocha-chrome-wrapper.

shaper commented on tools/mocha-chrome-wrapper:
```diff
@@ -0,0 +1,48 @@
+});
+Promise.all(promises).then((values) => {
+  let exitCode = values.reduce(
+    (accumulator, currentValue) => accumulator|=currentValue
```
nit: I think typical JS style prefers binary ops surrounded with spaces so `' |= '` (and same for string concats above), I think `prettier` does this for us so could try just running it on this file with the --single-quote param.

smalls replied:
I'll run prettier - I need to get a plugin so I can't forget.

shaper commented on package.json:
```diff
@@ -3,7 +3,7 @@
   "private": true,
   "scripts": {
-    "test": "mocha-chrome test/index.test.html"
+    "test": "tools/mocha-chrome-wrapper */test/index.test.html"
```
What you've done here seems reasonable to me. My thoughts on reading for reference:

I thought maybe `mocha` or `mocha-chrome` could sort this out for us (globbing html files, running them in a batch) but I guess it doesn't do anything fancy test-html-wise, all I see is this:

https://mochajs.org/#the-test-directory

So then we could debate putting `index.test.html` in the test directory vs. the one above it. What you have here seems preferred since we won't have to have the `test/` prefix on all of the files the html file references.

smalls replied:
I was also hoping that something else would do the globbing but I didn't find anything. When we do we should rip this out.

shaper commented on tools/mocha-chrome-wrapper:
```diff
@@ -0,0 +1,48 @@
+const fs = require('fs');
+
+let promises = process.argv.slice(2).map(testScript => {
+  // we're not doing a ton of checking for now, since we've relied on
```
nit: prefer capitalizing first word in comments so they read as a normal full sentence.

shaper commented on tools/mocha-chrome-wrapper:
```diff
@@ -0,0 +1,48 @@
+      resolve(stats.failures);
+    });
+    mochaChrome.on('failure', message => {
+      resolve(1);
```
Should we log the message or has mochaChrome handled that for us already? Same re: `stats` above.

smalls replied:
MochaChrome will log details (and those are all present).

shaper commented on tools/mocha-chrome-wrapper:
```diff
@@ -0,0 +1,48 @@
+  let exitCode = values.reduce(
+    (accumulator, currentValue) => accumulator|=currentValue
+  );
+  process.exit(exitCode);
```
For future consideration -- we might want logging here re: the url associated with this promise, to more quickly allow someone reading to see at process exit which of the html files failed if it's a non-zero exit code, but can wait to see what we think.

smalls replied:
Yeah this is good - it'll be much faster to just print this at the bottom.

smalls committed:
Prettier mocha-chrome-wrapper
prettier --single-quote --write mocha-chrome-wrapper

smalls committed:
print results at the end of the overall run
Most importantly, be specific about which URLs failed (or passed).
More specific logging is handled by mocha-chrome.

smalls commented:
Noop.
